observability/logging.go

```diff
@@ -0,0 +1,232 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package observability
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	binlogpb "google.golang.org/grpc/binarylog/grpc_binarylog_v1"
+	iblog "google.golang.org/grpc/internal/binarylog"
+	configpb "google.golang.org/grpc/observability/internal/config"
+	grpclogrecordpb "google.golang.org/grpc/observability/internal/logging"
+)
+
+// translateMetadata translates the metadata from Binary Logging format to
+// its GrpcLogRecord equivalent.
+func translateMetadata(m *binlogpb.Metadata) *grpclogrecordpb.GrpcLogRecord_Metadata {
+	var res grpclogrecordpb.GrpcLogRecord_Metadata
+	res.Entry = make([]*grpclogrecordpb.GrpcLogRecord_MetadataEntry, len(m.Entry))
+	for i, e := range m.Entry {
+		res.Entry[i] = &grpclogrecordpb.GrpcLogRecord_MetadataEntry{
+			Key:   e.Key,
+			Value: e.Value,
+		}
+	}
+	return &res
+}
+
+func setPeerIfPresent(binlogEntry *binlogpb.GrpcLogEntry, grpcLogRecord *grpclogrecordpb.GrpcLogRecord) {
+	if binlogEntry.GetPeer() != nil {
+		grpcLogRecord.PeerAddress = &grpclogrecordpb.GrpcLogRecord_Address{
+			Type:    grpclogrecordpb.GrpcLogRecord_Address_Type(binlogEntry.Peer.Type),
+			Address: binlogEntry.Peer.Address,
+			IpPort:  binlogEntry.Peer.IpPort,
+		}
+	}
+}
+
+type observabilityBinaryMethodLogger struct {
+	rpcID, serviceName, methodName string
+	originalMethodLogger           iblog.MethodLogger
+	wrappedMethodLogger            iblog.MethodLogger
+}
+
+func (ml *observabilityBinaryMethodLogger) Log(c iblog.LogEntryConfig) {
+	// Invoke the original MethodLogger to maintain backward compatibility
+	if ml.originalMethodLogger != nil {
+		ml.originalMethodLogger.Log(c)
+	}
+
+	// Fetch the compiled binary logging log entry
+	if ml.wrappedMethodLogger == nil {
+		logger.Info("No wrapped method logger found")
+		return
+	}
+	var binlogEntry *binlogpb.GrpcLogEntry
+	o, ok := ml.wrappedMethodLogger.(interface {
+		Build(iblog.LogEntryConfig) *binlogpb.GrpcLogEntry
+	})
+	if !ok {
+		logger.Errorf("Failed to locate the Build method in wrapped method logger")
+		return
+	}
+	binlogEntry = o.Build(c)
+
+	// Translate to GrpcLogRecord
+	grpcLogRecord := &grpclogrecordpb.GrpcLogRecord{
+		Timestamp:  binlogEntry.GetTimestamp(),
+		RpcId:      ml.rpcID,
+		SequenceId: binlogEntry.GetSequenceIdWithinCall(),
+		// Making DEBUG the default LogLevel
+		LogLevel: grpclogrecordpb.GrpcLogRecord_LOG_LEVEL_DEBUG,
+	}
+
+	switch binlogEntry.Logger {
+	case binlogpb.GrpcLogEntry_LOGGER_CLIENT:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_CLIENT
+	case binlogpb.GrpcLogEntry_LOGGER_SERVER:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_SERVER
+	default:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_UNKNOWN
+	}
```

Comment:
Consider using a map instead of a switch.  (Well, probably only if UNKNOWN is 0, which it should be.)